### PR TITLE
Integrate `device_vjp` with the QNode

### PR DESCRIPTION
**Pull request by albi3ro**
This PR allows users to request device-provided VJP's at the QNode level.

```pycon
>>> dev = qml.device('default.qubit')
>>> @qml.qnode(dev, diff_method="adjoint", device_vjp=True)
>>> def circuit(x):
...     qml.RX(x, wires=0)
...     return qml.expval(qml.PauliZ(0))
>>> with dev.tracker:
...     g = qml.grad(circuit)(qml.numpy.array(0.1))
>>> dev.tracker.totals
{'batches': 1, 'simulations': 1, 'executions': 1, 'vjp_batches': 1, 'vjps': 1}
>>> g
-0.09983341664682815
```

[sc-51039]

### CODE REVIEW

**Review by trbromley — COMMENTED** [review]
Thanks @albi3ro! This looks great. Happy to open this up to the team for review.

File: doc/releases/changelog-dev.md [hunk 1]
```diff
@@ -97,10 +97,23 @@
...
+```pycon
+>>> dev = qml.device('default.qubit')
+>>> @qml.qnode(dev, diff_method="adjoint", device_vjp=True)
+>>> def circuit(x):
+...     qml.RX(x, wires=0)
+...     return qml.expval(qml.PauliZ(0))
+>>> with dev.tracker:
+...     g = qml.grad(circuit)(qml.numpy.array(0.1))
+>>> dev.tracker.totals
+{'batches': 1, 'simulations': 1, 'executions': 1, 'vjp_batches': 1, 'vjps': 1}
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply doc/releases/changelog-dev.md:113]: A batch could contain more than one circuit. For example, broadcasting.

File: pennylane/qnode.py [hunk 2]
```diff
@@ -172,6 +172,8 @@ class QNode:
             the maximum number of derivatives to support. Increasing this value allows
             for higher order derivatives to be extracted, at the cost of additional
             (classical) computational overhead during the backwards pass.
+        device_vjp=False (bool): Whether or not to use the device provided Vector Jacobian Product (VJP).
+            A value of ``None`` indicates to use it if the device provides it, but use the full jacobian otherwise.
```
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/qnode.py:176]: Similar behaviour to `grad_on_execution`. `None` means "you can make the best choice".

[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/qnode.py:176]: > Similar behaviour to `grad_on_execution`. `None` means "you can make the best choice".

If this is the case why is the default not `None`?

[the remark this answers is not in the document]
Comment (mudit2812) [reply pennylane/qnode.py:176]: > Similar behaviour to `grad_on_execution`. `None` means "you can make the best choice".

If this is the case then doesn't it make more sense for the default to be `None` instead of `False`?

**Review by mudit2812 — APPROVED** [review]
Thanks Christina! What is expected to change internally and externally after `device_vjp` is fully integrated? This looks good!

**Comment by josh146** [comment]
A quick question: will we be able to make using VJPs internally the default for the user (without needing to specify a keyword argument) _if the underlying diff method, device, and interface support it_?

File: doc/releases/changelog-dev.md [hunk 3]
```diff
@@ -97,10 +97,23 @@
 
 * Autograd and torch can now use vjps provided by the device from the new device API. If a device provides
```
Comment (josh146) [inline doc/releases/changelog-dev.md:98]:
```suggestion
* Autograd and torch can now use VJPs provided by the device from the new device API. If a device provides
```

**Comment by albi3ro** [comment]
> A quick question: will we be able to make using VJPs internally the default for the user (without needing to specify a keyword argument) _if the underlying diff method, device, and interface support it_?

Once we have good support for VJP's we can switch the default from `False` to `None`. `None` will indicate to pennylane "use it if you can".  But we have a few things to add before then.

**Comment by albi3ro** [comment]
> Thanks Christina! What is expected to change internally and externally after `device_vjp` is fully integrated? This looks good!

You can see the new workflow in the code for autograd and torch (`interfaces/autograd.py` and `interfaces/torch.py`). PR #4878  is moving over the code for jax (non-jit).

File: pennylane/qnode.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (albi3ro) [reply pennylane/qnode.py:176]: Right now the best choice is rarely device vjps.  

PR #4841  will solve some of the performance issues, and make device vjps substantially faster than full jacobians for the gradient case, but `device_vjp` is still quite a bit slower for the full jacobian case.

We also don't support all interfaces, and even when we do add jax (non-jit) support in #4878, it will only support `jax.grad` and `jax.vjp`, not `jax.jacobian`.

**Review by timmysilv — APPROVED** [review]
